modules/loader/src/main/scala/com/snowplowanalytics/snowplow/postgres/source/PubSubEnv.scala

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2020-2021 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.postgres.source
+
+import cats.implicits._
+import cats.effect.{ContextShift, Blocker, Resource, Timer, ConcurrentEffect, Sync}
+
+import fs2.Pipe
+
+import org.log4s.getLogger
+
+import com.google.pubsub.v1.PubsubMessage
+
+import com.permutive.pubsub.consumer.grpc.{PubsubGoogleConsumerConfig, PubsubGoogleConsumer}
+import com.permutive.pubsub.consumer.Model.{ProjectId, Subscription}
+import com.permutive.pubsub.consumer.decoder.MessageDecoder
+import com.permutive.pubsub.consumer.ConsumerRecord
+
+import com.snowplowanalytics.snowplow.badrows.BadRow
+import com.snowplowanalytics.snowplow.postgres.config.LoaderConfig.Source
+import com.snowplowanalytics.snowplow.postgres.streaming.SinkPipe
+
+object PubSubEnv {
+
+  private lazy val logger = getLogger
+
+  def create[F[_]: ConcurrentEffect: ContextShift: Timer](blocker: Blocker, config: Source.PubSub): Resource[F, Environment[F, ConsumerRecord[F, String]]] = {
+    implicit val decoder: MessageDecoder[String] = pubsubDataDecoder
+    val project = ProjectId(config.projectId)
+    val subscription = Subscription(config.subscriptionId)
+    val pubsubConfig = PubsubGoogleConsumerConfig[F](onFailedTerminate = pubsubOnFailedTerminate[F])
+    Resource.eval(
+      ConcurrentEffect[F].delay(
+        Environment[F, ConsumerRecord[F, String]](
+          PubsubGoogleConsumer.subscribe[F, String](blocker, project, subscription, pubsubErrorHandler[F], pubsubConfig),
+          getPayload,
+          checkpointer,
+          SinkPipe.UnorderedPipe.forTransactor[F]
+        )
+      )
+    )
+  }
+
+  private def getPayload[F[_]](record: ConsumerRecord[F, String]): Either[BadRow, String] = record.value.asRight
+
+  private def checkpointer[F[_]]: Pipe[F, ConsumerRecord[F, String], Unit] = _.evalMap(_.ack)
```

Comment:
I think this can be `_.parEvalMapUnordered`.  And that _might_ improve throughput (although statements like that mean nothing until proven).  It might prevent acking from becoming a bottleneck.

For Kinesis apps we need to be more careful about checkpointing events in the correct order, to avoid data loss if the app crashes.  But for pubsub it is safe to ack events in any order.

Reply:
What should we give to `parEvalMapUnordered` for `maxConcurrent` ?

Reply:
As before, ideally it should be configurable. But in general, a good rule of thumb for everything `parSomething`-related is few hundreds. Something between 80 and 400, depending on machine's RAM and network.